packages/components/src/hooks/use-scroll-list-position.ts

```diff
@@ -0,0 +1,172 @@
+import { MutableRefObject, RefObject, useMemo, useRef } from 'react';
+import type { ListProps } from '../list/list';
+import type { ScrollListItemInfo, ScrollListProps } from '../scroll-list/scroll-list';
+import type { DimensionsById } from './use-element-rect';
+import { defaultPos, usePositionInParent } from './use-position';
+
+export interface ScrollListPositioningProps {
+    /**
+     * if set to false, items will not be unmounted when out of view
+     * @default true
+     */
+    unmountItems?: boolean;
+    /**
+     * Gap between items
+     *
+     * @default 0
+     */
+    itemGap?: number;
+    /**
+     * Items in a row or column
+     * @default 1
+     */
+    itemsInRow?: number;
+    /**
+     * Scroll offset if the scroll lists has a scroll window that is external to itself, it can have elements before it.
+     *
+     * For vertical lists, this affects scrollTop. For horizontal lists, this affects scrollLeft.
+     *   false: no remeasure,
+     *   true: measure on changes,
+     *   number: use the number as size
+     *
+     *   @default = 0
+     */
+    scrollOffset?: number | boolean;
+}
+
+export const useScrollListPosition = <T, EL extends HTMLElement>({
+    items,
+    getId,
+    getItemInfo,
+    itemSize,
+    isHorizontal,
+    unmountItems = true,
+    itemGap = 0,
+    itemsInRow = 1,
+    extraRenderSize = 0.5,
+    averageItemSize,
+    scrollWindowSize,
+    itemsDimensions,
+    scrollOffset = 0,
+    maxScrollSize,
+    scrollPosition,
+    scrollListRef,
+}: ScrollListPositioningProps & {
+    items: ListProps<T>['items'];
+    getId: ListProps<T>['getId'];
+    getItemInfo: (item: T) => ScrollListItemInfo<T>;
+    itemSize: ScrollListProps<T, EL>['itemSize'];
+    isHorizontal: boolean;
+    itemGap: ScrollListProps<T, EL>['itemGap'];
+    itemsInRow: ScrollListProps<T, EL>['itemsInRow'];
+    extraRenderSize: ScrollListProps<T, EL>['extraRenderSize'];
+    averageItemSize: number;
+    scrollWindowSize: number;
+    itemsDimensions: MutableRefObject<DimensionsById>;
+    maxScrollSize: number;
+    scrollPosition: number;
+    scrollListRef: RefObject<EL>;
+}) => {
+    const lastRenderedItem = useRef({
+        items,
+        last: 0,
+    });
+
+    if (lastRenderedItem.current.items !== items) {
+        lastRenderedItem.current.items = items;
+        lastRenderedItem.current.last = 0;
+    }
+
+    const shouldMeasureOffset = typeof scrollOffset === 'number' ? defaultPos : scrollOffset;
```

Comment:
use unmeasured instead of defaultPos or rename it to something more general and put it in `constants.ts`